### ISSUE TITLE
RFC 1: Arrays

### ISSUE THREAD

**Issue by kjnilsson** [kjnilsson/fez#15]
Arrays have no direct equivalent in `core erlang`. A lot of code that uses arrays can be quite directly translated to using lists instead (fsharp `List`s or `Seq`s). The exception to this is code that mutates the array directly. That said a lot of code may use arrays and it would be good to cover arrays in some shape or form.

The most direct equivalent of erlang's `binary()` type is the `byte array` so it would make sense to translate operations on byte arrays to operations on binaries. 

Arrays are generic so this leave a lot of types uncovered. Should the non mutable aspects and apis of `Array.*` that aren't byte arrays simply be translated as erlang lists? Are there any other options?

**Comment by kjnilsson** [comment]
Manual memory management just doesn't play nicely with the erlang vm typically.

**Comment by tsloughter** [comment]
Does it have to be a data structure available in core and not the stdlib? There is the `array` module, http://erlang.org/doc/man/array.html, which while not being an array like F#'s may give better performance when implementing than a list.

**Comment by kjnilsson** [comment]
@tsloughter yes I am leaning towards using the `array` module for any non-binary types with a `fez` specific `Array.setElement` function that returns the array. The problem is that fsharp arrays are mutable and there is no realistic means of supporting functions ( like `Array.set`) that return `unit`.  The gnarly bit of translating one language into another. :)

**Comment by tsloughter** [comment]
Ah yea, `set` on mutable needs to return `unit` :(. I'd be curious to a NIF like @rascala suggests, using resources to reference the array. But maybe garbage collection issues with that...

**Comment by kjnilsson** [comment]
Erlang (or rather hipe) has mutable arrays but only of immediate types. See
hipe_ bif

Probably not a practical approach though.
